File: cpp/source/actions/ConnectAction.cpp
#include "ConnectAction.hpp"

#include "Exception.hpp"

#include "AppAction.hpp"

#include "CalibrationSettings.hpp"

ConnectAction::ConnectAction(Listener* const _listener):
    ICommAction(_listener)
{
    state = IDLE;
    wasNonStatic = false;
}

void ConnectAction::start(void)
{
    monitor->trace("ConnectAction::start");
    state = INITIAL_COMMAND;
    sendSignal(SignalData::START_CMD, SignalData::START);
}

bool ConnectAction::isActionDone(void) const
{
    return state == IDLE;
}

ICommAction::Type ConnectAction::getType(void) const
{
    return CONNECT;
}

std::string ConnectAction::getStateName(void) const
{
    switch (state)
    {
    case IDLE: return "IDLE";
    case INITIAL_COMMAND: return "INITIAL_COMMAND";
    case CALIBRATION: return "CALIBRATION";
    case CALIBRATION_RECEPTION: return "CALIBRATION_RECEPTION";
    case FINAL_COMMAND: return "FINAL_COMMAND";
    default:
        __RL_EXCEPTION__("ConnectAction::getStateName::Unexpected state");
    }
}

void ConnectAction::handleReception(const IMessage& message)
{
    switch (state)
    {
    case CALIBRATION_RECEPTION:
        if (handleSignalPayloadReception(message))
        {
            state = FINAL_COMMAND;
            listener->send(SignalData(SignalData::APP_LOOP, SignalData::START));
            startSignalTimeout(SignalData::APP_LOOP);
            monitor->notifyUavEvent(new UavEventReceived(message));
        }
        break;

    default:
        except("Message received in unexpected state", message);
    }
}

void ConnectAction::handleSignalReception(const Parameter parameter)
{
    switch (state)
    {
    case INITIAL_COMMAND:
        switch (parameter)
        {
        case SignalData::ACK:
            monitor->trace("Initial command successfull");
            startSignalTimeout(SignalData::CALIBRATION_SETTINGS);
            state = CALIBRATION;
            break;

        default:
            except("Unexpected signal parameter message received", parameter);
        }
        break;

    case CALIBRATION:
        switch (parameter)
        {
        case SignalData::READY:
            monitor->trace("Calibration done");
            initializeSignalPayloadReception(SignalData::CALIBRATION_SETTINGS);
            state = CALIBRATION_RECEPTION;
            break;

        case SignalData::NON_STATIC:
            monitor->trace("Calibration non-static");
            startSignalTimeout(SignalData::CALIBRATION_SETTINGS);
            if (!wasNonStatic)
            {
                wasNonStatic = true;
                monitor->notifyUavEvent(new UavEvent(UavEvent::CALIBRATION_NON_STATIC));
            }
            break;

        default:
            except("Unexpected signal parameter received", parameter);
        }
        break;

    case FINAL_COMMAND:
        switch (parameter)
        {
        case SignalData::ACK:
            monitor->trace("App loop ready");
            state = IDLE;
            listener->startAction(new AppAction(listener));
            monitor->notifyUavEvent(new UavEvent(UavEvent::APPLICATION_LOOP_STARTED));
            break;

        default:
            except("Unexpected signal parameter message received", parameter);
        }
        break;

    default:
        except("Signal parameter in unexpected state", parameter);
    }
}

File: java/actions/ConnectAction.java
package com.skydive.java.actions;

import com.skydive.java.CommHandler;
import com.skydive.java.CommMessage;
import com.skydive.java.CommTask;
import com.skydive.java.UavEvent;
import com.skydive.java.data.CalibrationSettings;
import com.skydive.java.data.SignalData;
import com.skydive.java.events.CommEvent;
import com.skydive.java.events.SignalPayloadEvent;

import java.util.Timer;
import java.util.TimerTask;

/**
 * Created by Bartosz Nawrot on 2016-10-13.
 */
public class ConnectAction extends CommHandlerAction {

    private static final long CONNECTION_TIMEOUT = 5000;

    public enum ConnectState {
        IDLE,
        INITIAL_COMMAND,
        PROTOCOL_VERSION,
        WAITING_FOR_CALIBRATION,
        WAITING_FOR_CALIBRATION_DATA,
        FINAL_COMMAND,
    }

    private ConnectState state;

    private boolean connectionProcedureDone;

    private CommTask connectionTimeoutTask;

    public ConnectAction(CommHandler commHandler) {
        super(commHandler);
        state = ConnectState.IDLE;
        connectionProcedureDone = false;

        connectionTimeoutTask = new CommTask(1000.0 / CONNECTION_TIMEOUT) {
            @Override
            protected String getTaskName() {
                return "connection_timeout_task";
            }

            @Override
            protected void task() {
                if (state == ConnectState.INITIAL_COMMAND) {
                    ConnectAction.this.commHandler.getUavManager().notifyUavEvent(
                            new UavEvent(UavEvent.Type.ERROR,
                                    "Timeout waiting for initial command response."));
                }
            }
        };
    }

    @Override
    public boolean isActionDone() {
        return connectionProcedureDone;
    }

    @Override
    public void start() {
        System.out.println("Starting connection procedure");
        connectionProcedureDone = false;
        state = ConnectState.INITIAL_COMMAND;
        commHandler.send(new SignalData(SignalData.Command.START_CMD, SignalData.Parameter.START).getMessage());
        commHandler.startCommTask(connectionTimeoutTask);
    }

    @Override
    public void handleEvent(CommEvent event) throws Exception {
        ConnectState actualState = state;
        switch (state) {
            case INITIAL_COMMAND:
                if (event.matchSignalData(
                        new SignalData(SignalData.Command.START_CMD, SignalData.Parameter.ACK))) {
                    commHandler.stopCommTask(connectionTimeoutTask);
                    state = ConnectState.PROTOCOL_VERSION;
                    System.out.println("Initial command received successfully");
                } else {
                    System.out.println("Unexpected event received at state " + state.toString());
                }
                break;

            case PROTOCOL_VERSION:
                SignalData protocolResponse = event.getSignalData();
                if (protocolResponse.getCommand() == SignalData.Command.PROTOCOL_VERSION_VALUE &&
                        protocolResponse.getParameterValue() == CommMessage.PROTOCOL_VERSION) {
                    commHandler.send(new SignalData(SignalData.Command.PROTOCOL_VERSION, SignalData.Parameter.ACK).getMessage());
                    state = ConnectState.WAITING_FOR_CALIBRATION;
                    System.out.println("Protocol setup done");
                } else {
                    commHandler.send(new SignalData(SignalData.Command.PROTOCOL_VERSION, SignalData.Parameter.NOT_ALLOWED).getMessage());
                    commHandler.getUavManager().notifyUavEvent(new UavEvent(UavEvent.Type.ERROR, "Unsupported protocol version"));
                }
                break;

            case WAITING_FOR_CALIBRATION:
                if (event.matchSignalData(
                        new SignalData(SignalData.Command.CALIBRATION_SETTINGS, SignalData.Parameter.READY))) {
                    state = ConnectState.WAITING_FOR_CALIBRATION_DATA;
                    System.out.println("Calibration done successfully, data ready");
                } else if (event.matchSignalData(
                        new SignalData(SignalData.Command.CALIBRATION_SETTINGS, SignalData.Parameter.NON_STATIC))) {
                    System.out.println("Calibration non static");
                    commHandler.getUavManager().notifyUavEvent(new UavEvent(UavEvent.Type.CALIBRATION_NON_STATIC));
                } else {
                    System.out.println("Unexpected event received at state " + state.toString());
                }
                break;

            case WAITING_FOR_CALIBRATION_DATA:
                if (event.getType() == CommEvent.EventType.SIGNAL_PAYLOAD_RECEIVED
                        && ((SignalPayloadEvent) event).getDataType() == SignalData.Command.CALIBRATION_SETTINGS_DATA) {
                    SignalPayloadEvent signalEvent = (SignalPayloadEvent) event;

                    CalibrationSettings calibrationSettings = (CalibrationSettings) signalEvent.getData();
                    if (calibrationSettings.isValid()) {
                        System.out.println("Calibration settings received after adHoc calibration");
                        state = ConnectState.FINAL_COMMAND;
                        commHandler.send(new SignalData(SignalData.Command.CALIBRATION_SETTINGS, SignalData.Parameter.ACK).getMessage());
                        commHandler.getUavManager().setCalibrationSettings(calibrationSettings);
                        // send final start command
                        commHandler.send(new SignalData(SignalData.Command.APP_LOOP, SignalData.Parameter.START).getMessage());
                    } else {
                        System.out.println("Calibration settings received but the data is invalid, responding with DATA_INVALID");
                        commHandler.send(new SignalData(SignalData.Command.CALIBRATION_SETTINGS, SignalData.Parameter.DATA_INVALID).getMessage());
                    }
                } else {
                    System.out.println("Unexpected event received at state " + state.toString());
                }
                break;

            case FINAL_COMMAND:
                if (event.matchSignalData(
                        new SignalData(SignalData.Command.APP_LOOP, SignalData.Parameter.ACK))) {
                    connectionProcedureDone = true;
                    state = ConnectState.IDLE;
                    System.out.println("Final command received successfully, connection procedure done");
                    commHandler.getUavManager().notifyUavEvent(new UavEvent(UavEvent.Type.CONNECTED));
                } else {
                    System.out.println("Unexpected event received at state " + state.toString());
                }
                break;

            default:
                throw new Exception("Event: " + event.toString() + " received at unknown state");
        }
        if (actualState != state) {
            System.out.println("HandleEvent done, transition: " + actualState.toString() + " -> " + state.toString());
        } else {
            System.out.println("HandleEvent done, no state change");
        }
    }

    @Override
    public ActionType getActionType() {
        return ActionType.CONNECT;
    }
}